File: Leetcode/0128-LongestConsecutiveSequence/cpp/solution2.cpp
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  int longestConsecutive(vector<int>& nums) {
    if (nums.empty()) {
      return 0;
    }

    unordered_set<int> set(nums.begin(), nums.end());
    int longestLength = 1;
    for (int num : set) {
      if (set.contains(num - 1)) {
        continue;
      }
      int length = 1;
      int next = num + 1;
      while (set.contains(next)) {
        next++;
        length++;
      }
      longestLength = max(longestLength, length);
    }
    return longestLength;
  }
};

File: Leetcode/0128-LongestConsecutiveSequence/java/Solution2.java
import java.util.HashSet;

public class Solution2 {
    public int longestConsecutive(int[] nums) {
        // Edge case
        if (nums.length == 0) {
            return 0;
        }

        HashSet<Integer> set = new HashSet<>();
        for (int n : nums) {
            set.add(n);
        }
        int longest = 0;
        for (int n : set) {
            if (set.contains(n-1)) {
                continue;
            }
            int i = n+1;
            while (set.contains(i)) {
                i++;
            }
            int length = i-n;
            longest = Math.max(longest, length);
        }
        return longest;
    }
}
